notebooks/Mitigating-Disparities-in-Ranking-from-Binary-Data.ipynb

```diff
@@ -0,0 +1,592 @@
+    "classifier_summary.loc['disparity']=(classifier_summary.loc['white']-classifier_summary.loc['black']).abs()\n",
+    "classifier_summary.loc['disparity', classifier_summary.columns.str.startswith('auc')]='-'\n",
+    "display(classifier_summary)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We see that exponentiated gradient variants generally achieve lower disparity on this data than `ThresholdOptimizer`. Without access to the sensitive feature at the test time, this comes at the cost of bringing the AUC essentially to that of a random coin toss (AUC of **expgrad_X** is close to 0.5). With access to the sensitive feature, the overall AUC is comparable to that achieved by `ThresholdOptimizer`, but `ThresholdOptimizer` achieves a better worst-case AUC across the two population. \n",
+    "\n",
+    "*Note*: `ExponentiatedGradient` produces randomized predictions (similarly to `ThresholdOptimizer`), so the AUC and selection rate will vary if you re-run the cell above. Also, because of a mismatch between the training and test distributions and because of finite samples, we do not expect `ExponentiatedGradient` to achieve test disparity equal to 0.01.\n",
+    "\n",
+    "We next show that if we are willing to tolerate a larger demographic disparity, it is possible to achieve non-trivial AUC values even without access to the sensitive feature. We run `ExponentiatedGradient` with the bound on the training disparity equal to 0.3:"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "expgrad_X_alt = ExponentiatedGradient(\n",
+    "    LogisticRegression(solver='liblinear', fit_intercept=True),\n",
+    "    constraints=DemographicParity(),\n",
+    "    eps=0.3, nu=1e-6)\n",
+    "expgrad_X_alt.fit(\n",
+    "    X_train.iloc[balanced_index,:],\n",
+    "    y_train.iloc[balanced_index],\n",
+    "    sensitive_features=A_train.iloc[balanced_index])\n",
+    "scores_expgrad_X_alt = pd.Series(\n",
+    "    expgrad_X.predict(X_test), name=\"scores_expgrad_X_alt\")\n",
+    "auc_expgrad_X_alt = group_metric_as_df(\n",
+    "    \"auc_expgrad_X_alt\",\n",
+    "    group_roc_auc_score(y_test, scores_expgrad_X_alt, A_test))\n",
+    "sel_expgrad_X_alt = group_metric_as_df(\n",
+    "    \"selection_expgrad_X_alt\",\n",
+    "    group_mean_prediction(y_test, scores_expgrad_X_alt, A_test))\n",
+    "auc_expgrad_X_alt.loc['disparity'] = '-'\n",
+    "sel_expgrad_X_alt.loc['disparity'] = (sel_expgrad_X_alt.loc['white'] - sel_expgrad_X_alt.loc['black']).abs()\n",
+    "\n",
+    "display(pd.concat([auc_expgrad_X_alt, sel_expgrad_X_alt], axis=1))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Comparing Classifiers using the Dashboard\n",
+    "\n",
+    "We finish this section by comparing the four predictors above using `FairlearnDashboard`:"
```

Comment:
Might be helpful to highlight a couple of things to look at. I suppose it depends on whether we're trying to make this notebook teach people just about `fairlearn` or about ML fairness in general.

Reply:
Let's do that in a future PR?